src/main/scala/components/csr.scala

```diff
@@ -0,0 +1,533 @@
+  val reg_mtval = Reg(UInt(32.W))
+  val reg_mscratch = Reg(UInt(32.W))
+  val reg_mtimecmp = Reg(UInt(64.W))
+  val reg_medeleg = Reg(UInt(32.W))
+
+  val reg_mip = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_mie = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_wfi = RegInit(false.B)
+  val reg_mtvec = RegInit(0.U.asTypeOf(new MTVec()))
+
+  val reg_time = WideCounter(64)
+  val reg_instret = WideCounter(64, io.retire)
+
+  val reg_mcounterinhibit = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+  val reg_mcounteren = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+
+  val read_mstatus = io.status.asUInt()
+  val isa_string = "I"
+  val reg_misa = RegInit((BigInt(0) | isa_string.map(x => 1 << (x - 'A')).reduce(_|_)).U.asTypeOf(new MISA()))
+  val reg_mvendorid = RegInit(0.U.asTypeOf(new MVendorID()))
+
+
+  val read_mapping = collection.mutable.LinkedHashMap[Int,Bits](
+    MCSRs.mcounterinhibit -> reg_mcounterinhibit.asUInt, 
+    MCSRs.mcycle -> reg_time,
+    MCSRs.minstret -> reg_instret,
+    MCSRs.mimpid -> 0.U,
+    MCSRs.marchid -> 0.U,
+    MCSRs.mvendorid -> 0.U,
+    MCSRs.misa -> reg_misa.asUInt,
+    MCSRs.mstatus -> read_mstatus,
+    MCSRs.mtvec -> MCSRCmd.TRAPADDR,
+    MCSRs.mip -> reg_mip.asUInt(),
+    MCSRs.mie -> reg_mie.asUInt(),
+    MCSRs.mscratch -> reg_mscratch,
+    MCSRs.mepc -> reg_mepc,
+    MCSRs.mtval -> reg_mtval,
+    MCSRs.mcause -> reg_mcause.asUInt(),
+    MCSRs.mhartid -> 0.U,
+    MCSRs.medeleg -> reg_medeleg)
+    
+  read_mapping += MCSRs.mcycleh -> 0.U 
+  read_mapping += MCSRs.minstreth -> 0.U
+  
+  //CSR DECODE
+  val cmd = WireInit(3.U(3.W))
+  
+  when( io.decode.inst(6, 0) === ("b1110011".U)){
+    switch(io.decode.inst(14, 12)){
+      is("b011".U){
+        cmd := MCSRCmd.clear
+        io.regwrite := true.B
+      }
+      is("b111".U){
+        cmd := MCSRCmd.clear
+        io.regwrite := true.B
+      } 
+      is("b010".U){
+        cmd := MCSRCmd.set
+        io.regwrite := true.B
+      }
+      is("b110".U){
+        cmd := MCSRCmd.set
+        io.regwrite := true.B
+      }
+      is("b001".U){
+        cmd := MCSRCmd.write
+        io.regwrite := true.B
+      }
+      is("b101".U){
+        cmd := MCSRCmd.write
+        io.regwrite := true.B
+      }
+      is("b000".U){
+        cmd := MCSRCmd.interrupt
+        io.regwrite := false.B
+      }
+    }
+  }.otherwise{
+    cmd := MCSRCmd.nop
+    io.regwrite := false.B
+  }
+  
+  val csr = io.decode.inst(MCSRCmd.MSB, MCSRCmd.LSB)
+  val system_insn = cmd === MCSRCmd.interrupt
+  val cpu_ren = cmd =/= MCSRCmd.nop && !system_insn
+
+
+  val decoded_addr = read_mapping map { case (k, v) => k -> (csr === k) }
+  val priv_sufficient = MCSRCmd.MPRV >= csr(9,8)
+  val read_only = csr(11,10).andR
+  val cpu_wen = cpu_ren && cmd =/= MCSRCmd.read && priv_sufficient
+  val wen = cpu_wen && !read_only
+  val wdata = readModifyWriteCSR(cmd.asInstanceOf[UInt], io.rw.rdata, io.rw.wdata)
+
+  val opcode = 1.U << csr(2,0)
```

Comment:
This is weird code... Do you think it's a hardware optimization?

I was going to ask you to replace with a `when` statement, but now that I think about it maybe we shouldn't re-write too much until we have solid tests in place so we don't break things.

Reply:
Im not sure if it helps with optimization. I guess I'll leave it alone for now, but add a comment next to it